tests/unit-tests/test-class-sensei-course-enrolment.php

```diff
@@ -0,0 +1,385 @@
+	}
+
+	/**
+	 * Tests to make sure non-handling providers aren't used. Never_Handles has a true response, but it should never be asked.
+	 */
+	public function testEnrolmentCheckVersionCachingWorks() {
+		$course_id  = $this->getSimpleCourse();
+		$student_id = $this->createStandardStudent();
+
+		// Initial request.
+		$this->resetAndSetUpVersiondProvider( false ); // Version : 1; Odd numbers do not provide enrolment.
+		$course_enrolment_a = Sensei_Course_Enrolment::get_course_instance( $course_id );
+		$is_enroled_a       = $course_enrolment_a->is_enroled( $student_id );
+		$is_enroled_a_2     = $course_enrolment_a->is_enroled( $student_id );
+		$this->assertFalse( $is_enroled_a, 'This version of the provider should not provide enrolment' );
+		$this->assertEquals( $is_enroled_a_2, $is_enroled_a, 'Duplicate calls should use caching and return same result' );
+
+		// Bump version. Second request.
+		$this->resetAndSetUpVersiondProvider( true ); // Version : 2; Even numbers provide enrolment.
+		$course_enrolment_b = Sensei_Course_Enrolment::get_course_instance( $course_id );
+		$is_enroled_b       = $course_enrolment_b->is_enroled( $student_id );
+		$this->assertTrue( $is_enroled_b, 'This version of the provider should provide enrolment. Cache was not refreshed on version change' );
+	}
+
+	/**
+	 * Tests a provider that actually has an opinion about which students are enroled. Only enrols users with "Dinosaur" in their name.
+	 */
+	public function testEnrolmentCheckConditionalUserPositive() {
+		$course_id           = $this->getSimpleCourse();
+		$student_id_standard = $this->createStandardStudent();
+		$student_id_dino     = $this->createDinosaurStudent();
+
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Provides_For_Dinosaurs::class );
+
+		$course_enrolment = Sensei_Course_Enrolment::get_course_instance( $course_id );
+
+		$this->assertFalse( $course_enrolment->is_enroled( $student_id_standard ), 'This provider should only allow dinosaurs' );
+		$this->assertTrue( $course_enrolment->is_enroled( $student_id_dino ), 'This provider should allow dinosaurs' );
+	}
+
+	/**
+	 * Tests a provider that actually has an opinion about which students are enroled. Denies crooks.
+	 */
+	public function testEnrolmentCheckConditionalUserNegative() {
+		$course_id        = $this->getSimpleCourse();
+		$student_id_crook = $this->turnStudentIntoCrook( $this->createStandardStudent() );
+		$student_id_okay  = $this->createStandardStudent();
+
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Denies_Crooks::class );
+
+		$course_enrolment = Sensei_Course_Enrolment::get_course_instance( $course_id );
+
+		$this->assertFalse( $course_enrolment->is_enroled( $student_id_crook ), 'This provider should deny crooks' );
+		$this->assertTrue( $course_enrolment->is_enroled( $student_id_okay ), 'This provider should allow okay students' );
+	}
+
+	/**
+	 * Helper for test above.
```

Comment:
Should we move this method 2 methods up?

Reply:
0c80b60fbf0e34a8e5eb5c2c4124ba00f7021d0e